Mathlib/SetTheory/Ordinal/Veblen.lean

```diff
@@ -0,0 +1,332 @@
+  (veblenWith_right_strictMono hf o).le_iff_le
+
+theorem veblenWith_injective (o : Ordinal) : Function.Injective (veblenWith f o) :=
+  (veblenWith_right_strictMono hf o).injective
+
+@[simp]
+theorem veblenWith_inj : veblenWith f o a = veblenWith f o b ↔ a = b :=
+  (veblenWith_injective hf o).eq_iff
+
+theorem right_le_veblenWith (o a : Ordinal) : a ≤ veblenWith f o a :=
+  (veblenWith_right_strictMono hf o).le_apply
+
+theorem veblenWith_left_monotone (o : Ordinal) : Monotone (veblenWith f · o) := by
+  rw [monotone_iff_forall_lt]
+  intro a b h
+  rw [← veblenWith_veblenWith_of_lt hf h]
+  exact (veblenWith_right_strictMono hf a).monotone (right_le_veblenWith hf b o)
+
+theorem veblenWith_pos (hp : 0 < f 0) (o a : Ordinal) : 0 < veblenWith f o a := by
+  have H (b) : 0 < veblenWith f 0 b := by
+    rw [veblenWith_zero]
+    exact hp.trans_le (hf.monotone (Ordinal.zero_le _))
+  obtain rfl | h := Ordinal.eq_zero_or_pos o
+  · exact H a
+  · rw [← veblenWith_veblenWith_of_lt hf h]
+    exact H _
+
+theorem veblenWith_zero_strictMono (hp : 0 < f 0) : StrictMono (veblenWith f · 0) := by
+  intro a b h
+  dsimp only
+  rw [← veblenWith_veblenWith_of_lt hf h, veblenWith_lt_veblenWith_iff_right hf]
+  exact veblenWith_pos hf hp b 0
+
+theorem veblenWith_zero_lt_veblenWith_zero (hp : 0 < f 0) :
+    veblenWith f o₁ 0 < veblenWith f o₂ 0 ↔ o₁ < o₂ :=
+  (veblenWith_zero_strictMono hf hp).lt_iff_lt
+
+theorem veblenWith_zero_le_veblenWith_zero (hp : 0 < f 0) :
+    veblenWith f o₁ 0 ≤ veblenWith f o₂ 0 ↔ o₁ ≤ o₂ :=
+  (veblenWith_zero_strictMono hf hp).le_iff_le
+
+theorem veblenWith_zero_inj (hp : 0 < f 0) : veblenWith f a 0 = veblenWith f b 0 ↔ a = b :=
+  (veblenWith_zero_strictMono hf hp).injective.eq_iff
+
+theorem left_le_veblenWith (hp : 0 < f 0) (a b : Ordinal) : a ≤ veblenWith f a b :=
+  (veblenWith_zero_strictMono hf hp).le_apply.trans <|
+    (veblenWith_right_strictMono hf _).monotone (Ordinal.zero_le _)
+
+theorem IsNormal.veblenWith_zero (hp : 0 < f 0) : IsNormal (veblenWith f · 0) := by
+  rw [isNormal_iff_strictMono_limit]
+  refine ⟨veblenWith_zero_strictMono hf hp, fun o ho a IH ↦ ?_⟩
+  rw [veblenWith_of_ne_zero f ho.pos.ne', derivFamily_zero]
+  apply nfpFamily_le fun l ↦ ?_
+  suffices ∃ b < o, List.foldr _ 0 l ≤ veblenWith f b 0 by
+    obtain ⟨b, hb, hb'⟩ := this
+    exact hb'.trans (IH b hb)
+  induction l with
+  | nil => use 0; simp [ho.pos]
+  | cons a l IH =>
+    obtain ⟨b, hb, hb'⟩ := IH
+    refine ⟨_, ho.succ_lt (max_lt a.2 hb), ((veblenWith_right_strictMono hf _).monotone <|
+      hb'.trans <| veblenWith_left_monotone hf _ <|
+        (le_max_right a.1 b).trans (Order.le_succ _)).trans ?_⟩
+    rw [veblenWith_veblenWith_of_lt hf]
+    rw [Order.lt_succ_iff]
+    exact le_max_left _ b
+
+theorem cmp_veblenWith_eq : cmp (veblenWith f o₁ a) (veblenWith f o₂ b) =
+  match cmp o₁ o₂ with
+    | .eq => cmp a b
+    | .lt => cmp a (veblenWith f o₂ b)
+    | .gt => cmp (veblenWith f o₁ a) b := by
```

Comment:
Please start multiline return types on a new line. Also no need to indent extra the branches of a `match`
```suggestion
theorem cmp_veblenWith_eq :
    cmp (veblenWith f o₁ a) (veblenWith f o₂ b) =
      match cmp o₁ o₂ with
      | .eq => cmp a b
      | .lt => cmp a (veblenWith f o₂ b)
      | .gt => cmp (veblenWith f o₁ a) b := by
```